datalake/src/main/java/com/sequenceiq/datalake/flow/upgrade/ccm/UpgradeCcmActions.java

```diff
@@ -0,0 +1,127 @@
+package com.sequenceiq.datalake.flow.upgrade.ccm;
+
+import static com.sequenceiq.datalake.flow.upgrade.ccm.UpgradeCcmStateSelectors.UPGRADE_CCM_FAILED_HANDLED_EVENT;
+import static com.sequenceiq.datalake.flow.upgrade.ccm.UpgradeCcmStateSelectors.UPGRADE_CCM_FINALIZED_EVENT;
+
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.statemachine.StateContext;
+import org.springframework.statemachine.action.Action;
+
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.datalake.entity.DatalakeStatusEnum;
+import com.sequenceiq.datalake.entity.SdxCluster;
+import com.sequenceiq.datalake.flow.SdxContext;
+import com.sequenceiq.datalake.flow.upgrade.ccm.event.UpgradeCcmFailedEvent;
+import com.sequenceiq.datalake.flow.upgrade.ccm.event.UpgradeCcmStackEvent;
+import com.sequenceiq.datalake.flow.upgrade.ccm.event.UpgradeCcmStackRequest;
+import com.sequenceiq.datalake.flow.upgrade.ccm.event.UpgradeCcmSuccessEvent;
+import com.sequenceiq.datalake.metric.MetricType;
+import com.sequenceiq.datalake.metric.SdxMetricService;
+import com.sequenceiq.datalake.service.AbstractSdxAction;
+import com.sequenceiq.datalake.service.sdx.status.SdxStatusService;
+import com.sequenceiq.flow.core.FlowEvent;
+import com.sequenceiq.flow.core.FlowParameters;
+import com.sequenceiq.flow.core.FlowState;
+
+@Configuration
+public class UpgradeCcmActions {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UpgradeCcmActions.class);
+
+    @Inject
+    private SdxStatusService sdxStatusService;
+
+    @Inject
+    private SdxMetricService metricService;
+
+    @Bean(name = "UPGRADE_CCM_UPGRADE_STACK_STATE")
+    public Action<?, ?> upgradeCcmStack() {
+        return new AbstractSdxAction<>(UpgradeCcmStackEvent.class) {
+
+            @Override
+            protected SdxContext createFlowContext(FlowParameters flowParameters, StateContext<FlowState, FlowEvent> stateContext,
+                    UpgradeCcmStackEvent payload) {
+                return SdxContext.from(flowParameters, payload);
+            }
```

Comment:
Nit: As this overridden implementation seems to be copy-pasted in all the anonymous action classes, I suggest creating an `abstract` subclass of `AbstractSdxAction` (e.g. `AbstractUpgradeCcmAction`) that contains this override, and then all the anonymous action instances can be a subclass of `AbstractUpgradeCcmAction` without having to override `createFlowContext()`. `AbstractUpgradeCcmAction` can be a private nested (`static`) class in `UpgradeCcmActions`, and have a type parameter like `<E extends SdxEvent>`.

Besides `createFlowContext()`, `AbstractUpgradeCcmAction` can also implement `getFailurePayload()` as the trivial `return payload;`, so that it need not be implemented explicitly. (This particular subclass in `upgradeCcmStack()` will still have to override `getFailurePayload()`, though.)

Reply:
done